5931 maven/gradle project setup (#234)
Xceptance/xlt-documentation: content/en/xlt/load-testing/Advanced/210-gradle-builds.md

### Plugins
Include the following in your `build.gradle` file to use Gradle's Java plugin and its provided build steps:

```groovy
plugins {
    id 'java'
}
```

### Adding Repository Location
XLT is published to [Maven Central](https://search.maven.org/artifact/com.xceptance/xlt). To integrate XLT into your Gradle project, copy and paste the following into your `build.gradle` file:

```groovy
/* Add Maven Central repository information */

repositories {
    maven {
        url = uri('https://repo.maven.apache.org/maven2/')
    }
}
```

### Dependency Scopes
XLT is provided at runtime by the container, so it does not need to be packaged with your project build, thus reducing the upload size when starting a load test. Maven has the `provided` scope for dependencies that are needed in your classpath at compile time but not at runtime; however since there is no such scope defined in Gradle you have to define it yourself in your `build.gradle` file: 

```groovy
configurations {
    provided
}

sourceSets {
    main {
         // Add dependencies of type 'provided' to compile classpath
        compileClasspath += configurations.provided
    }
}
```  

> 

Reviewer: In case your testsuite's code is organized in a different way, for example when your XLT tests resides in `src/test/java`, you may have to update the compile classpath of the proper [source set](https://docs.gradle.org/7.4.2/userguide/java_plugin.html#source_sets) as well.

### Adding XLT to your Project
We can then add xlt as a provided dependency:

```groovy
dependencies {
    provided 'com.xceptance:xlt:8.2.0'
}
```  

{{% note title="Version Update" %}}
When configuring your test project to use a newer XLT version, do not forget to update XLT on your load machines as well. The version you’ve used to develop your test scripts must match the execution version of your load test environment.
{{% /note %}}

### Copying Dependencies

> If your test suite is uses external dependencies or libraries, they must be copied into the test suite as part of the compile or package step. XLT does not build the project on the agent machines and therefore does not resolve dependencies there. It simply uploads the test suite to the agent, including the contents of the `build` directory.

Reviewer: ```suggestion
If your test suite makes use of any external dependencies or libraries, they have to be copied to a proper location within you test suite where XLT can find them, ideally as part of the compile or package step. XLT does not build the project on the agent machines and therefore does not resolve dependencies there. It simply uploads the test suite to the agent, including the contents of the `build` directory.
```

To automatically copy all non-provided dependencies to `build/dependency` at compile time, add the following snippet to your `build.gradle`:

```groovy
tasks.register('copyDeps', Copy) {
  into layout.buildDirectory.dir('dependency')
  from configurations.testRuntimeClasspath, configurations.runtimeClasspath
}

tasks.withType(JavaCompile) {

>     options.encoding = 'UTF-8'

Reviewer: Nope, they are important as they configure the Java compiler (more specifically, the source file encoding and JavaSE release to use).

>     options.encoding = 'UTF-8'

Author: I mean they _are_ important for the build.gradle file, but are not necessarily the same for any user setting up his testsuite using gradle, so probably they should not be included in this howto?

>     options.encoding = 'UTF-8'

Reviewer: We may include these settings as comments such that people know that they can configure the Java compiler options here. WDYT?

>     options.encoding = 'UTF-8'

Author: I am fine with leaving them as they are, with comments which part does what. Should be clear enough then.

>     options.encoding = 'UTF-8'

Reviewer: Okay. I'd also be fine with the statements commented out. :shrug:

>     options.encoding = 'UTF-8'
>     options.release = 11

Reviewer: ```suggestion
    // Configure Java compiler (source file encoding and JavaSE release)
    options.encoding = 'UTF-8'
    options.release = 11
```

>     dependsOn('copyDeps')

Reviewer: ```suggestion
    // Depend on 'copyDeps' task such that dependencies are copied automatically
    dependsOn('copyDeps')
```
